### PR TITLE
snapctl: don't error out on start/stop/restart from configure hook during install or refresh

### PR DESCRIPTION
**Pull request by stolowski**
Check if snapctl start/stop/restart is run as part of same change when checking for conflicting tasks and don't report the conflict if so.

### CODE REVIEW

File: overlord/servicestate/servicestate.go [hunk 1]
```diff
@@ -39,7 +39,7 @@ type Instruction struct {
 
 type ServiceActionConflictError struct{ error }
 
-func Control(st *state.State, appInfos []*snap.AppInfo, inst *Instruction) (*state.TaskSet, error) {
+func Control(st *state.State, appInfos []*snap.AppInfo, inst *Instruction, conflictCheck func(*state.Task) bool) (*state.TaskSet, error) {
```
Comment (pedronis) [inline overlord/servicestate/servicestate.go]: I would make this take a context and not the predicate function

File: overlord/hookstate/ctlcmd/helpers.go [hunk 2]
```diff
@@ -96,7 +96,19 @@ func runServiceCommand(context *hookstate.Context, inst *servicestate.Instructio
 		return err
 	}
 
-	ts, err := servicestateControl(st, appInfos, inst)
+	isSameChange := func(otherTask *state.Task) bool {
+		if context.IsEphemeral() {
+			return false
+		}
+		task, ok := context.Task()
```
Comment (pedronis) [inline overlord/hookstate/ctlcmd/helpers.go]: no reason to do this prep bits inside the function, they could be outside, ex. if context is ephemeral don't pass a function, pass see comment below, this helper probably doesn't belong here (at least until we have a strong case)

File: overlord/servicestate/servicestate.go (hunk 1 above)
Comment (stolowski) [reply overlord/servicestate/servicestate.go]: Done

File: overlord/hookstate/ctlcmd/helpers.go [hunk 3]
```diff
@@ -69,20 +69,21 @@ func getServiceInfos(st *state.State, snapName string, serviceNames []string) ([
...
 	if !ok {
-		panic("attempted to queue command with ephemeral context")
+		return fmt.Errorf("attempted to queue command with ephemeral context")
 	}
-	change := task.Change()
-	tasks := change.Tasks()
+
+	change := hookTask.Change()
+	tasks := change.LaneTasks(hookTask.Lanes()...)
 	ts.WaitAll(state.NewTaskSet(tasks...))
```
Comment (pedronis) [inline overlord/hookstate/ctlcmd/helpers.go]: I think we also want to have all the tasks in ts to join the same lanes as hooTask, unless those are just [0]

File: overlord/servicestate/servicestate.go [hunk 4]
```diff
@@ -83,7 +84,21 @@ func Control(st *state.State, appInfos []*snap.AppInfo, inst *Instruction) (*sta
 
 	st.Lock()
 	defer st.Unlock()
-	if err := snapstate.CheckChangeConflictMany(st, snapNames, nil); err != nil {
+
+	var checkConflict func(otherTask *state.Task) bool
+	if context != nil && !context.IsEphemeral() {
+		if task, ok := context.Task(); ok {
+			checkConflict = func(otherTask *state.Task) bool {
+				if task.Change() != nil && otherTask.Change() != nil {
```
Comment (pedronis) [inline overlord/servicestate/servicestate.go]: taks.Change could be gotten outside the function

File: overlord/hookstate/ctlcmd/helpers.go (hunk 3 above)
Comment (pedronis) [reply overlord/hookstate/ctlcmd/helpers.go]: do we need extra unit tests for this bit?

**Comment by stolowski** [comment]
Unfortunately there seems to be some kind of problem after I switched to use lanes (I think that's when the problem started). My spread test reaches kill-timeout on restart in the last check where configure hook restarts two services on install. `snap change <ID>` shows that the entire install change is waiting on exactly these two restarts to finish. For some reason this takes very long time (a couple of minutes) and reaches killtimeout of the test. The change however completes successfuly after a couple of minutes (I've verified that in the spread test shell).

File: overlord/hookstate/ctlcmd/helpers.go [hunk 5]
```diff
@@ -70,16 +70,21 @@ func getServiceInfos(st *state.State, snapName string, serviceNames []string) ([
...
 
+	st := context.State()
+	st.Lock()
+	defer st.Unlock()
+
 	change := hookTask.Change()
-	tasks := change.LaneTasks(hookTask.Lanes()...)
+	hookTaskLanes := hookTask.Lanes()
+	tasks := change.LaneTasks(hookTaskLanes...)
+	if hookTaskLanes[0] != 0 {
```
Comment (pedronis) [inline overlord/hookstate/ctlcmd/helpers.go]: we should also check that len()==1

File: overlord/hookstate/ctlcmd/helpers.go [hunk 6]
```diff
@@ -70,16 +70,21 @@ func getServiceInfos(st *state.State, snapName string, serviceNames []string) ([
...
+		ts.JoinLane(hookTaskLanes[0])
```
Comment (pedronis) [inline overlord/hookstate/ctlcmd/helpers.go]: though it's usually one we should probably join all of them

**Comment by pedronis** [comment]
something sounds very odd about the restart taking so long? also not sure how lanes would play a role, unless the task waiting we setup is wrong, lanes in themselves are only relevant on abort

File: overlord/hookstate/ctlcmd/helpers.go [hunk 7]
```diff
@@ -69,20 +69,31 @@ func getServiceInfos(st *state.State, snapName string, serviceNames []string) ([
...
+	} else {
+		for _, l := range hookTaskLanes {
+			ts.JoinLane(l)
+		}
 	}
-	change := task.Change()
-	tasks := change.Tasks()
 	ts.WaitAll(state.NewTaskSet(tasks...))
 	change.AddAll(ts)
+	st.EnsureBefore(0)
```
Comment (pedronis) [inline overlord/hookstate/ctlcmd/helpers.go]: ah, good catch! this probably merits a comment, maybe something like:

// as this can be run from what was originally the last task of a change, make sure the tasks added to the change are considered immediately

Comment (stolowski) [reply overlord/hookstate/ctlcmd/helpers.go]: Comment added.

File: overlord/hookstate/ctlcmd/helpers.go [hunk 8]
```diff
@@ -69,20 +69,32 @@ func getServiceInfos(st *state.State, snapName string, serviceNames []string) ([
...
 	st.Lock()
 	defer st.Unlock()
 
-	task, ok := context.Task()
-	if !ok {
-		panic("attempted to queue command with ephemeral context")
+	change := hookTask.Change()
+	hookTaskLanes := hookTask.Lanes()
+	tasks := change.LaneTasks(hookTaskLanes...)
+	if len(hookTaskLanes) == 1 && hookTaskLanes[0] != 0 {
```
Comment (pedronis) [inline overlord/hookstate/ctlcmd/helpers.go]: the if/else seems wrong, if the lanes are {0}  we don't need/want to join anything, maybe something like:

    if len(hookTaskLanes) == 1 && hooTaskLanes[0] == 0 {
       hookTaskLanes := nil
    }

    for _, l := range hookTaskLanes{
    ...

File: overlord/hookstate/ctlcmd/helpers.go [hunk 9]
```diff
@@ -69,20 +69,32 @@ func getServiceInfos(st *state.State, snapName string, serviceNames []string) ([
...
+	} else {
+		for _, l := range hookTaskLanes {
+			ts.JoinLane(l)
+		}
 	}
-	change := task.Change()
-	tasks := change.Tasks()
 	ts.WaitAll(state.NewTaskSet(tasks...))
 	change.AddAll(ts)
+	// as this can be run from what was originally the last task of a change, make sure the tasks added to the change are considered immediately.
```
Comment (pedronis) [inline overlord/hookstate/ctlcmd/helpers.go]: seems to need wrapping

File: overlord/snapstate/snapstate.go [hunk 10]
```diff
@@ -536,6 +536,7 @@ func InstallMany(st *state.State, names []string, userID int) ([]string, []*stat
 			return nil, nil, err
 		}
 		installed = append(installed, name)
+		ts.JoinLane(st.NewLane())
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go]: do we have a test for this?

File: overlord/hookstate/ctlcmd/helpers.go (hunk 8 above)
Comment (pedronis) [reply overlord/hookstate/ctlcmd/helpers.go]: we probably need a test for the no lanes aka lanes {0} case

File: overlord/servicestate/servicestate.go [hunk 11]
```diff
@@ -83,7 +84,22 @@ func Control(st *state.State, appInfos []*snap.AppInfo, inst *Instruction) (*sta
 
 	st.Lock()
 	defer st.Unlock()
-	if err := snapstate.CheckChangeConflictMany(st, snapNames, nil); err != nil {
+
+	var checkConflict func(otherTask *state.Task) bool
+	if context != nil && !context.IsEphemeral() {
+		if task, ok := context.Task(); ok {
+			chg := task.Change()
+			checkConflict = func(otherTask *state.Task) bool {
```
Comment (pedronis) [inline overlord/servicestate/servicestate.go]: do we need unit tests for this logic?

File: overlord/snapstate/snapstate_test.go [hunk 12]
```diff
@@ -6321,8 +6321,12 @@ func (s *snapmgrTestSuite) TestInstallMany(c *C) {
 	c.Assert(tts, HasLen, 2)
 	c.Check(installed, DeepEquals, []string{"one", "two"})
 
-	for _, ts := range tts {
+	for i, ts := range tts {
 		verifyInstallTasks(c, 0, 0, ts, s.state)
+		// check that tasksets are in separate lanes
+		for _, t := range ts.Tasks() {
+			c.Assert(t.Lanes(), DeepEquals, []int{i + 1})
```
Comment (pedronis) [inline overlord/snapstate/snapstate_test.go]: I just noticed that Update* didn't have a test about the JoinLane in doUpdate either, maybe worth a drive-by addition

File: overlord/hookstate/ctlcmd/helpers.go [hunk 13]
```diff
@@ -69,20 +69,32 @@ func getServiceInfos(st *state.State, snapName string, serviceNames []string) ([
...
 	st.Lock()
 	defer st.Unlock()
 
-	task, ok := context.Task()
-	if !ok {
-		panic("attempted to queue command with ephemeral context")
+	change := hookTask.Change()
+	hookTaskLanes := hookTask.Lanes()
+	tasks := change.LaneTasks(hookTaskLanes...)
+	if len(hookTaskLanes) == 1 && hookTaskLanes[0] == 0 {
```
Comment (zyga) [inline overlord/hookstate/ctlcmd/helpers.go]: I'm a bit lost here. Can you please document the function and the purpose of joining those lanes?

File: overlord/hookstate/ctlcmd/helpers.go [hunk 14]
```diff
@@ -69,20 +69,32 @@ func getServiceInfos(st *state.State, snapName string, serviceNames []string) ([
...
+	for _, l := range hookTaskLanes {
+		ts.JoinLane(l)
 	}
-	change := task.Change()
-	tasks := change.Tasks()
 	ts.WaitAll(state.NewTaskSet(tasks...))
 	change.AddAll(ts)
+	// As this can be run from what was originally the last task of a change,
+	// make sure the tasks added to the change are considered immediately.
+	st.EnsureBefore(0)
```
Comment (zyga) [inline overlord/hookstate/ctlcmd/helpers.go]: I'm not sure this is something that belongs in queueCommand, it feels like unexpected side-effect.

File: overlord/hookstate/ctlcmd/services_test.go [hunk 15]
```diff
@@ -95,6 +130,8 @@ func (s *servicectlSuite) SetUpTest(c *C) {
 	s.st.Lock()
 	defer s.st.Unlock()
 
+	snapstate.ReplaceStore(s.st, &s.fakeStore)
```
Comment (zyga) [inline overlord/hookstate/ctlcmd/services_test.go]: Does this need to be undone?

File: overlord/hookstate/ctlcmd/services_test.go [hunk 16]
```diff
@@ -228,14 +265,139 @@ func (s *servicectlSuite) TestRestartCommand(c *C) {
...
+	c.Check(installed, DeepEquals, []string{"one", "two"})
+	c.Assert(tts, HasLen, 2)
+	c.Assert(tts[0].Tasks(), HasLen, 12)
+	c.Assert(tts[1].Tasks(), HasLen, 12)
+	chg.AddAll(tts[0])
+	chg.AddAll(tts[1])
+
+	s.st.Unlock()
+
+	for i := 0; i < 2; i++ {
```
Comment (zyga) [inline overlord/hookstate/ctlcmd/services_test.go]: Nitpick, you can just iterate over tts's directly.

File: overlord/hookstate/ctlcmd/services_test.go [hunk 17]
```diff
@@ -228,14 +265,139 @@ func (s *servicectlSuite) TestRestartCommand(c *C) {
...
+		_, _, err = ctlcmd.Run(context, []string{"start", "test-snap.test-service"})
+		c.Check(err, IsNil)
+		_, _, err = ctlcmd.Run(context, []string{"restart", "test-snap.test-service"})
+		c.Check(err, IsNil)
+	}
+
+	s.st.Lock()
+	defer s.st.Unlock()
+
+	for i := 1; i < 3; i++ {
```
Comment (zyga) [inline overlord/hookstate/ctlcmd/services_test.go]: IMO `i <= 2` would be more natural here.

File: overlord/hookstate/ctlcmd/services_test.go [hunk 18]
```diff
@@ -228,14 +265,139 @@ func (s *servicectlSuite) TestRestartCommand(c *C) {
...
+	c.Check(installed, DeepEquals, []string{"test-snap", "other-snap"})
+	c.Assert(tts, HasLen, 2)
+	c.Assert(tts[0].Tasks(), HasLen, 16)
+	c.Assert(tts[1].Tasks(), HasLen, 16)
+	chg.AddAll(tts[0])
+	chg.AddAll(tts[1])
+
+	s.st.Unlock()
+
+	for i := 0; i < 2; i++ {
```
Comment (zyga) [inline overlord/hookstate/ctlcmd/services_test.go]: Same as above.

File: overlord/hookstate/ctlcmd/services_test.go [hunk 19]
```diff
@@ -228,14 +265,139 @@ func (s *servicectlSuite) TestRestartCommand(c *C) {
...
+		_, _, err = ctlcmd.Run(context, []string{"start", "test-snap.test-service"})
+		c.Check(err, IsNil)
+		_, _, err = ctlcmd.Run(context, []string{"restart", "test-snap.test-service"})
+		c.Check(err, IsNil)
+	}
+
+	s.st.Lock()
+	defer s.st.Unlock()
+
+	for i := 1; i < 3; i++ {
```
Comment (zyga) [inline overlord/hookstate/ctlcmd/services_test.go]: Same as above.

File: tests/lib/snaps/test-snapd-service/meta/hooks/configure [hunk 20]
```diff
@@ -6,9 +6,10 @@ COMMAND=$(snapctl get command)
 if [ "$COMMAND" != "" ]; then
     if [ "$COMMAND" = "restart" ]; then
         snapctl restart test-snapd-service.test-snapd-service
-        snapctl restart test-snapd-service.test-snapd-other-service
     else
         snapctl "$COMMAND" test-snapd-service.test-snapd-service
     fi
 fi
 sleep 3
+
```
Comment (zyga) [inline tests/lib/snaps/test-snapd-service/meta/hooks/configure]: Can you document why this is significant so that it doesn't go away by accident later.

File: overlord/hookstate/ctlcmd/restart.go [hunk 21]
```diff
@@ -33,8 +33,7 @@ The restart command restarts the given services of the snap. If executed from th
 )
 
 func init() {
-	// FIXME: uncomment once the feature is fixed to work on install/refresh
-	// addCommand("restart", shortRestartHelp, longRestartHelp, func() command { return &restartCommand{} })
+	addCommand("restart", shortRestartHelp, longRestartHelp, func() command { return &restartCommand{} })
```
Comment (zyga) [inline overlord/hookstate/ctlcmd/restart.go]: Just a comment, I find it curious we disabled this at the client, not at the daemon.

File: overlord/snapstate/snapstate.go (hunk 10 above)
Comment (zyga) [reply overlord/snapstate/snapstate.go]: @stolowski can you please add a comment why this is relevant:
```
Handle each installation in a separate lane so that unrelated snaps can install successfully in face of failure to install a particular one.
```

CC: @mvo5 will this be a problem for your prerequisite work?

File: overlord/snapstate/snapstate_test.go [hunk 22]
```diff
@@ -599,6 +599,11 @@ func (s *snapmgrTestSuite) TestUpdateMany(c *C) {
 
 	ts := tts[0]
 	verifyUpdateTasks(c, unlinkBefore|cleanupAfter, 3, ts, s.state)
+
+	// check that the tasks are in non-default lane
+	for _, t := range ts.Tasks() {
+		c.Assert(t.Lanes(), DeepEquals, []int{1})
```
Comment (zyga) [inline overlord/snapstate/snapstate_test.go]: Should each task be in the same lane? I would expect that each install is in separate one.

File: overlord/hookstate/ctlcmd/restart.go (hunk 21 above)
Comment (stolowski) [reply overlord/hookstate/ctlcmd/restart.go]: No, this *is* the daemon side. snapctl is the only place where the entire commands, including the help are generated on the daemon side and the client sends a very simple json with argv, literally. The client is implemented in cmd/snapctl.

File: overlord/snapstate/snapstate.go (hunk 10 above)
Comment (mvo5) [reply overlord/snapstate/snapstate.go]: @zyga It might have funny effects, afaict if the snap foo gets instaleld in a lane and it requires a prereq that prereq will get installed in the default lane. Which means that if foo fails the prereq is still instaleld. or worse if the prereq fails foo also gets installed. I think we need code in the prepreq install that puts the prereq into the same lane.

But stepping back, this change does not really fit into the title of the PR, maybe a separate PR so that its more clear what change is happening here.

Comment (pedronis) [reply overlord/snapstate/snapstate.go]: no, the prereqs are put in their own lane, not the default lane (that was intentional afair)

File: overlord/hookstate/ctlcmd/helpers.go (hunk 13 above)
Comment (pedronis) [reply overlord/hookstate/ctlcmd/helpers.go]: we want the additional tasks to be in the same lane(s) as the original hook

File: overlord/servicestate/servicestate.go [hunk 23]
```diff
@@ -39,7 +40,7 @@ type Instruction struct {
 
 type ServiceActionConflictError struct{ error }
 
-func Control(st *state.State, appInfos []*snap.AppInfo, inst *Instruction) (*state.TaskSet, error) {
+func Control(st *state.State, appInfos []*snap.AppInfo, inst *Instruction, context *hookstate.Context) (*state.TaskSet, error) {
```
Comment (pedronis) [inline overlord/servicestate/servicestate.go]: this doesn't have a doc comment, it's probably merits one so that we can also explain context

File: overlord/hookstate/ctlcmd/helpers.go [hunk 24]
```diff
@@ -96,14 +108,13 @@ func runServiceCommand(context *hookstate.Context, inst *servicestate.Instructio
 		return err
 	}
 
-	ts, err := servicestateControl(st, appInfos, inst)
+	ts, err := servicestateControl(st, appInfos, inst, context)
```
Comment (pedronis) [inline overlord/hookstate/ctlcmd/helpers.go]: probably merits a comment like:

// passing context so we can ignore self-conflicts with the current change

**Comment by pedronis** [comment]
Question: where is the part that actually allows us to run service commands while installation is in progress?

@zyga it's the bits that pass down context and then consider current Change so that we can ignore self-conflicts in the conflict checking logic

File: overlord/snapstate/snapstate_test.go (hunk 22 above)
Comment (stolowski) [reply overlord/snapstate/snapstate_test.go]: This test addition is to address previous comment from @pedronis about missing test for already existing case with lanes in updateMany test. And it's ok since this test actually updates a single snap, so we see a single (non-default) lane. I'll however move this change to a separate PR.

File: overlord/hookstate/ctlcmd/helpers.go (hunk 14 above)
Comment (stolowski) [reply overlord/hookstate/ctlcmd/helpers.go]: The is needed because we are adding task(s) after the last task that has already been processed. Not having this causes a very long pause at runtime. We do similiar thing in a couple of different places. I'm not sure if there is a better place to do this.

Comment (zyga) [reply overlord/hookstate/ctlcmd/helpers.go]: If @pedronis agrees then +1

File: overlord/hookstate/ctlcmd/services_test.go (hunk 15 above)
Comment (stolowski) [reply overlord/hookstate/ctlcmd/services_test.go]: I don't think we need to. None of the tests we currently have restore it after calling ReplaceStore.

File: overlord/snapstate/snapstate.go (hunk 10 above)
Comment (stolowski) [reply overlord/snapstate/snapstate.go]: @mvo Ok, I'll move this change (plus a tiny change to UpdateMany test) into a separate PR.

File: overlord/hookstate/ctlcmd/services_test.go (hunk 15 above)
Comment (pedronis) [reply overlord/hookstate/ctlcmd/services_test.go]: the store is attached to the cache in the state, each test creates a new state, so no

**Comment by pedronis** [comment]
seems there's a real issue with order that isn't stable in one of the tests:

services_test.go:345:
    c.Check(installed, DeepEquals, []string{"test-snap", "other-snap"})
... obtained []string = []string{"other-snap", "test-snap"}
... expected []string = []string{"test-snap", "other-snap"}
